src/nupic/research/temporal_memory.py

```diff
@@ -340,7 +341,12 @@ def learnOnSegments(self,
     @param predictedInactiveCells       (set)         Indices of predicted inactive cells
     @param prevMatchingSegments         (set)         Indices of segments with
     """
-    for segment in prevActiveSegments | learningSegments:
+    # Sort segments before iterating for compatibility with C++
+    segments = prevActiveSegments | learningSegments
+    segments = sorted(list(segments),
+                      key=lambda segment: connections.cellForSegment(segment))
```

Comment:
Can multiple segments belong to the same cell? If so, then we may need a secondary key here to make sure segments for the same cell have a strict ordering.